src/PollySandbox/Extensions/ResilienceExtensions.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Martin Costello, 2023. All rights reserved.
+// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
+
+using System.Net;
+using System.Reflection;
+using System.Text;
+using System.Threading.RateLimiting;
+using Polly;
+using Polly.CircuitBreaker;
+using Polly.Extensions.Telemetry;
+using Polly.RateLimiting;
+using Polly.Registry;
+using Polly.Retry;
+using Polly.Telemetry;
+using Polly.Timeout;
+using Refit;
+
+namespace PollySandbox.Extensions;
+
+public static partial class ResilienceExtensions
+{
+    public static IServiceCollection AddResilience(this IServiceCollection services)
+    {
+        return AddResilience(services, new Dictionary<string, IList<string>>
+        {
+            ["Movies"] = GetEndpointNames<MoviesClient>(),
+            ["Users"] = GetEndpointNames<UsersClient>(),
+        });
+    }
+
+    public static IServiceCollection AddResilience(this IServiceCollection services, IDictionary<string, IList<string>> endpoints)
+    {
+        services.AddOptions<TelemetryOptions>()
+                .Configure<IMetricsPublisher>(
+                    (options, publisher)
+                        => options.OnTelemetryEvent = (args) => OnTelemetry(publisher, args));
+
+        foreach ((var name, var resources) in endpoints)
```

Comment:
Do you really need to setup all possible combination of resilience strategies right away?

Looking at the code you have this setup:

- Rate limitier (applied per endpoint, partitioned)
- Timeout (stateless, per endpoint)
- Circuit Breaker - applied per each endpoint + resource
- Retry - stateless

You could introduce `public readonly record struct ResourceEndpoint(string BuilderName, string Endpoint, string Resource)` complex key and with just a single `AddResilienceStrategy` call you could setup the whole pipeline dynamically.

Then on the consuming side just provide a key that identifies the resource/endpoint you want to call and the registry can create and cache the resilience strategy for that endpoint + resource automatically. 

The only difficulty here is the rate limiter which is applied to a whole endpoint. But you can create it as this:

``` csharp
// add resilience strategy, keyed by EndpointKey that only defines the builder name
services.AddResilienceStrategy(new EndpointKey("endpoint-pipeline", string.Empty, string.Empty, new()), (builder, context) =>
{
    var endpointOptions = context.StrategyKey.EndpointOptions;
    var registry = context.ServiceProvider.GetRequiredService<ResilienceStrategyRegistry<string>>();

    // we want to limit the number of concurrent requests per endpoint and not include the resource.
    // using a registry we can create and cache the shared resilience strategy
    var rateLimiterStrategy = registry.GetOrAddStrategy($"rate-limiter/{context.StrategyKey.EndpointName}", b =>
    {
        b.AddConcurrencyLimiter(new ConcurrencyLimiterOptions { PermitLimit = endpointOptions.MaxParallelization });
    });

    builder.AddStrategy(rateLimiterStrategy);

   // configure more strategies
}
```

I have also updated the sample here:
https://github.com/App-vNext/Polly/pull/1366

Reply:
My rationale was that re-using the common policies in an endpoint for each resource was better than having an identically configured one for each of the N resources. Do you think it's better to have that duplication and have just N endpoints' worth of strategies in the registry instead?

Reply:
> Do you think it's better to have that duplication and have just N endpoints' worth of strategies in the registry instead?

It would be problem for stateful strategies (RateLimiter, CircuitBreaker) and if you would be recreating them on a hot path or having 10000s of them. All other strategies are very cheap so you can duplicate them without problems. 

Also each endpoint might have different configuration so even when having many of them, some of these have different configuration for each endpoint.